Handles custom tab sizes for gist embeds and shortcodes

ChaosExAnima opened the pull request:
Fixes #10534

#### Changes proposed in this Pull Request:

* This adds functionality for tab sizing for gist embeds and shortcodes via `?ts=` query param.
* Adds `ts` attribute to shortcodes.

**Known limitation:** Currently, AMP [does not](https://amp.dev/documentation/examples/components/amp-gist/) have a way to change tab sizing. This is a limitation of the AMP `amp-gist` element.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This adds functionality to the embed/shortcode module.

#### Testing instructions:

* Enable Shortcode Embeds module.
* Create a post with a gist embed that appends `?ts=4` to the gist URL.
* Create a post with a gist shortcode that has `ts=4` in it.

Screenshot: 

![Screenshot on 2019-10-22 at 17_05_54](https://user-images.githubusercontent.com/426388/67367134-31621100-f575-11e9-80cc-5d23b785b1b7.png)

#### Proposed changelog entry for your changes:

* Adds support for tab sizing for gist embeds.

ChaosExAnima committed:
[not verified] Handles custom tab sizes for gist embeds and shortcodes

ChaosExAnima commented on modules/shortcodes/gist.php:
```diff
@@ -156,6 +166,12 @@ function github_gist_shortcode( $atts, $content = '' ) {
 
+	// Set the tab size, allowing attributes to override the query string.
+	$tab_size = $gist_info['ts'];
+	if ( ! empty( $atts['ts'] ) ) {
```
I decided to allow the attributes to override the query string to allow for theme filtering via [`shortcode_atts_{$shortcode}`](https://developer.wordpress.org/reference/hooks/shortcode_atts_shortcode/) and similar.

jeherve replied:
Should we add some default attributes in `shortcode_atts` at the top of `github_gist_shortcode` to allow for this to happen?

ChaosExAnima replied:
The code already sets the defaults in `jetpack_gist_get_shortcode_id`. Do you feel that it needs to have that set additionally?

ChaosExAnima commented on modules/shortcodes/gist.php:
```diff
@@ -195,7 +211,11 @@ class_exists( 'Jetpack_AMP_Support' )
 	// inline style to prevent the bottom margin to the embed that themes like TwentyTen, et al., add to tables.
-	$return = '<style>.gist table { margin-bottom: 0; }</style><div class="gist-oembed" data-gist="' . esc_attr( $id ) . '"></div>';
+	$return = sprintf(
+		'<style>.gist table { margin-bottom: 0; }</style><div class="gist-oembed" data-gist="%1$s" data-ts="%2$d"></div>',
```
Noting that we embed this multiple times. May be a better way to do this.

jeherve commented on modules/shortcodes/gist.php:
```diff
@@ -58,6 +58,7 @@ function jetpack_gist_get_shortcode_id( $gist = '' ) {
 	$gist_info = array(
 		'id'   => '',
 		'file' => '',
+		'ts'   => 8,
```
I noticed you set the default to 8 here, and to 4 in `github_gist_simple_embed`. What do you think about having the same default everywhere? 4 seems like a sensible choice as 8 may push content a bit too much on sites with a narrow content width.

ChaosExAnima replied:
I meant to be consistent, can adjust the function below. I chose 8 as that's the GitHub gist default.

ChaosExAnima replied:
Fixed in `55d9440abcade22d64e60b490889c8641500c4e6`

jeherve reviewed (changes requested):
What do you think about adding a new Unit Test to cover this use-case to our existing Gist tests here:
`tests/php/modules/shortcodes/test-class.gist.php`

ChaosExAnima committed:
Makes tab size default consistent

ChaosExAnima committed:
Fixes tests to no longer fail

Resolves bugs with older tests that were failing with tab spacing code.

ChaosExAnima committed:
[not verified] Fixes incorrect test and slash trimming method.

Test had extraneous whitespace, which let to WP parsing `/]` as closing of tag. Changed method of slash trimming to use `trim` and fixed test.

ChaosExAnima committed:
[not verified] Ensures default tab size is set if function exits early

ChaosExAnima committed:
[not verified] Adds tests to cover tab size functionality

ChaosExAnima commented:
@jeherve - Did a bit of tinkering, and it looks like adding `shortcode_atts` will actually break everything as the code uses `empty` a lot instead of setting defaults. I'd suggest releasing this iteration now and making a cleanup ticket to rectify the issues there?

jeherve commented:
That works for me!
